src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,132 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int NATURAL_NUMBER_BORDERLINE = 0;
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double CAPACITY_MULTIPLY_BY = 1.5;
+    private static final String NEGATIVE_INDEX_ERROR = "Index should be natural number";
+    private static final String INDEX_OUT_OF_SIZE_ERROR = "Index out of list length";
+
+    private T[] data;
+    private int size;
+
+    public ArrayList() {
+        data = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int capacity) {
+        if (capacity > 0) {
+            data = (T[]) new Object[capacity];
+        } else if (capacity == 0) {
+            data = (T[]) new Object[]{};
+        } else {
+            throw new ArrayListIndexOutOfBoundsException(NEGATIVE_INDEX_ERROR);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (size < data.length) {
+            data[size] = value;
+            size++;
+        } else {
+            this.ensureCapacity();
+            this.add(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        isIndexValid(index);
+        if (size < data.length) {
+            System.arraycopy(data, index, data, index + 1, size - index);
+            data[index] = value;
+            size++;
+        } else {
+            this.ensureCapacity();
+            this.add(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size + list.size() > data.length) {
+            this.ensureCapacity();
+            this.addAll(list);
+        }
+        for (int i = size; i < (size + list.size()); i++) {
+            data[i] = list.get(i - size);
+        }
+        size += list.size();
     }
```

Comment:
this can be done much more easily with a single loop.

Reply:
I hope you mean that the if() check is redundant. Looks like it is.

Because size update is important

Reply:
```
for (int i = 0; i < list.size(); i++) {
            add(list.get(i));
        }
```
that's all what u need

Reply:
Wow, it really easier, than mine :`(